wagtail/contrib/redirects/views.py

```diff
@@ -126,3 +133,183 @@ def add(request):
     return TemplateResponse(request, "wagtailredirects/add.html", {
         'form': form,
     })
+
+
+@permission_checker.require_any("add")
+def start_import(request):
+    from_encoding = "utf-8"
+    query_string = request.GET.get('q', "")
+
+    if not request.POST:
+        SUPPORTED_FORMATS = ["CSV", "TSV", "XLS", "XLSX"]
+        accepted_formats = [
+            x for x in DEFAULT_FORMATS if x.__name__ in SUPPORTED_FORMATS
+        ]
+        return render(
+            request,
+            "wagtailredirects/choose_import_file.html",
+            {
+                'search_form': SearchForm(
+                    data=dict(q=query_string) if query_string else None, placeholder=_("Search redirects")
+                ),
+                "form": ImportForm(accepted_formats),
+            },
+        )
+
+    form_kwargs = {}
+    form = ImportForm(
+        DEFAULT_FORMATS, request.POST or None, request.FILES or None, **form_kwargs
+    )
+
+    if not form.is_valid():
+        return render(
+            request,
+            "wagtailredirects/choose_import_file.html", {
+                'search_form': SearchForm(
+                    data=dict(q=query_string) if query_string else None, placeholder=_("Search redirects")
+                ),
+                "form": form,
+            }
+        )
+
+    import_formats = get_import_formats()
+    input_format = import_formats[int(form.cleaned_data["input_format"])]()
+    import_file = form.cleaned_data["import_file"]
+    file_storage = write_to_file_storage(import_file, input_format)
+
+    try:
+        data = file_storage.read(input_format.get_read_mode())
+        if not input_format.is_binary() and from_encoding:
+            data = force_str(data, from_encoding)
+        dataset = input_format.create_dataset(data)
+    except UnicodeDecodeError as e:
+        return HttpResponse(_(u"<h1>Imported file has a wrong encoding: %s</h1>" % e))
+    except Exception as e:  # pragma: no cover
+        return HttpResponse(
```

Comment:
Is it possible for both of these error responses to use the response.messages system and land you back on the import page with the errors in red at the top of the page? Being shown these full page errors, outside of the Wagtail interface is a bit heavy.

Reply:
@lb- Good point! Its definitely doable. Will solve in my next update.

Reply:
Also, not sure if response.messages is the best, maybe non-field form errors.